lib/pbench/server/api/__init__.py

```diff
@@ -75,6 +99,156 @@ def create_app():
     return app
 
 
+def _build_cors_preflight_response():
+    response = make_response()
+    response = _corsify_actual_response(response)
+    response.headers.add("Access-Control-Allow-Headers", "*")
+    response.headers.add("Access-Control-Allow-Methods", "*")
+    return response
+
+
+def _corsify_actual_response(response):
+    response.headers.add("Access-Control-Allow-Origin", "*")
+    return response
+
+
+class CorsOptionsRequest:
+    """
+    From the client side, an OPTIONS request is initiated before every POST to request CORS.
+    On the server side, the server must specify the response headers to grant CORS accordingly.
+    """
+
+    def __init__(self, options_exception_msg):
+        self.options_exception_msg = options_exception_msg
+
+    def options(self):
+        try:
+            response = _build_cors_preflight_response()
+        except Exception:
+            app.logger.exception(self.options_exception_msg)
+            abort(400, message=self.options_exception_msg)
+        response.status_code = 200
+        return response
+
+
+class GraphQL(Resource, CorsOptionsRequest):
+    """GraphQL API for post request via server."""
+
+    def __init__(self):
+        CorsOptionsRequest.__init__(self, "Bad options request for the GraphQL query")
+
+    def post(self):
+        global app
+        graphQL = f"{app.config_graphql.get('host')}:{app.config_graphql.get('port')}"
+        json_data = request.get_json()
```

Comment:
According to the `flask` [documentation](https://flask.palletsprojects.com/en/1.1.x/api/#flask.Request.on_json_loading_failed), a `BadRequest` error will be raised if the JSON fails to parse.  Don't we have to catch that error and handle it properly?

Reply:
Maybe there should be some `flask` api annotations around this? That could centralize the behavior instead of duplicating the `except`/`abort` code across the various APIs...

https://flask-restplus.readthedocs.io/en/stable/errors.html

Reply:
Yes, however, I chose to pass the parameter `silent=True` to `get_json`, it will silence the `BadRequest` exception and return None instead, if it fails to parse the json. According to documentation ` on_json_loading_failed()` will only get called if the mimetype fails to indicate if its a json. So most of the time if its not a json it will return None, and by passing `silent=True` the parsing exception will also return None. therefore both conditions can be handled by checking if return value is None or not

Reply:
Not sure I follow ... is it true that as it stands right now, `get_json()` will never raise a `BadRequest` exception?

Reply:
Yes, if parsing fails `get_json(silent=True)` will return None instead of calling `on_json_loading_failed()` (`on_json_loading_failed()` raises the BadRequest if it couldt return a value) [doc](https://flask.palletsprojects.com/en/1.1.x/api/#flask.Request.get_json)